### PR TITLE
Add SSH jump host support

### PR DESCRIPTION
**Pull request by adisbladis**
This work is based on https://github.com/NixOS/nixops/pull/1270 because they touch a lot of the same code.

Additionally as a requisite for this work I have changed the "TCP ping" to actually log in to the remote host and run a command instead of just checking the TCP port availability.

Closes https://github.com/NixOS/nixops/issues/1150
Closes https://github.com/NixOS/nixops/issues/784

### CODE REVIEW

**Comment by adisbladis** [comment]
Rebased on master since #1270 was merged.

File: nixops/backends/__init__.py [hunk 1]
```diff
@@ -247,16 +247,29 @@ def reboot_sync(self, hard: bool = False) -> None:
         """Reboot this machine and wait until it's up again."""
         self.reboot(hard=hard)
         self.log_start("waiting for the machine to finish rebooting...")
-        nixops.util.wait_for_tcp_port(
```
Comment (grahamc) [inline nixops/backends/__init__.py:250]: Should we explicitly add a deprecation message to this, or even delete the function?

File: nixops/backends/__init__.py [hunk 2]
```diff
@@ -247,16 +247,29 @@ def reboot_sync(self, hard: bool = False) -> None:
...
-            callback=lambda: self.log_continue("."),
+
+        nixops.util.wait_for_fail(
+            fn=lambda: self.ssh.run_command(
+                ["true"],
+                user=self.ssh_user,
+                forward_agent=False,
+                logged=False,
+                connection_tries=1,
+            )
```
Comment (grahamc) [inline nixops/backends/__init__.py]: Let's wrap this in a convenience method of like `is the machine actually reachable?`

File: nixops/backends/__init__.py [hunk 3]
```diff
@@ -247,16 +247,29 @@ def reboot_sync(self, hard: bool = False) -> None:
         """Reboot this machine and wait until it's up again."""
         self.reboot(hard=hard)
         self.log_start("waiting for the machine to finish rebooting...")
-        nixops.util.wait_for_tcp_port(
-            self.get_ssh_name(),
-            self.ssh_port,
-            open=False,
-            callback=lambda: self.log_continue("."),
```
Comment (grahamc) [inline nixops/backends/__init__.py:254]: Can we get the `.`s back? I think they are pretty important from a "is nixops even still trying, or is it stuck?" perspective.

File: nixops/backends/none.py [hunk 4]
```diff
@@ -109,6 +109,7 @@ def _check(self, res):
             res.exists = False
             return
         res.exists = True
+        # TODO: Use a better method that works with bastion hosts
         res.is_up = nixops.util.ping_tcp_port(self.target_host, self.ssh_port)
```
Comment (grahamc) [inline nixops/backends/none.py:113]: I think we're guaranteed to have a working SSH, so let's just use the same "ssh and run `true`" strategy as in other places. @aanderse mentioned some backends might not need to use SSH, but that is a big enough thing to be a whole feature request and "epic".

File: nixops/util.py [hunk 5]
```diff
@@ -398,26 +398,50 @@ def ping_tcp_port(
...
 
 
-def wait_for_tcp_port(
-    ip: str,
-    port: int,
-    timeout: int = -1,
-    open: bool = True,
-    callback: Optional[Callable[[], Any]] = None,
+def wait_for_success(
+    fn: Callable, timeout: int = -1, callback: Optional[Callable[[], Any]] = None,
```
Comment (grahamc) [inline nixops/util.py:402]:
```suggestion
    fn: Callable, timeout: Optional[int] = None, callback: Optional[Callable[[], Any]] = None,
```

Let's use our type system to our advantage here, and in the next function too.

**Comment by grahamc** [comment]
Since we're opting to go for more of a user-supplied "use sshOptions to do it", we'll need to bolster this use case by writing a user's guide on how to deploy with a jump host.

File: nixops/backends/__init__.py (hunk 3 above)
Comment (lovesegfault) [reply nixops/backends/__init__.py:254]: Can there be a real spinner as opposed to just dots? A quick google points me to https://github.com/manrajgrover/py-spinners and https://github.com/verigak/progress/

Comment (adisbladis) [reply nixops/backends/__init__.py:276]: For now I just re-added the dots. I'd love to make the output more pretty, but that's a separate PR.

**Pull request merged**